mlflow/genai/utils/trace_utils.py

```diff
@@ -394,32 +394,200 @@ def _to_dict(obj: Any) -> dict[str, Any]:
+            try:
+                contexts = [_parse_chunk(chunk) for chunk in retrieval_span.outputs or []]
+                retrieved[retrieval_span.span_id] = [c for c in contexts if c is not None]
+            except Exception as e:
+                _logger.debug(
+                    f"Fail to get retrieval context from span: {retrieval_span}. Error: {e!r}"
+                )
+
+        if retrieved:
+            return retrieved
+
+    return _try_extract_retrieval_context_with_llm(trace, model)
+
+
+def _try_extract_retrieval_context_with_llm(
+    trace: Trace,
+    model: str | None = None,
+) -> dict[str, list[dict[str, Any]]]:
+    """
+    Use LLM with tools to extract retrieval context from trace.
+
+    This is a fallback when programmatic extraction fails. The LLM uses tool calling
+    to examine trace spans and locate retrieval contexts.
+
+    Args:
+        trace: The MLflow trace object to analyze.
+        model: Optional model URI to use for extraction. If None, uses default model.
+
+    Returns:
+        Dictionary mapping span IDs to lists of retrieved chunks, in the same format
+        as extract_retrieval_context_from_trace.
+    """
+    from mlflow.genai.judges.constants import _DATABRICKS_DEFAULT_JUDGE_MODEL
+    from mlflow.genai.judges.utils import (
+        get_chat_completions_with_structured_output,
+        get_default_model,
+    )
+    from mlflow.types.llm import ChatMessage
+
+    output_example = json.dumps(
+        RetrievedChunksForTrace(
+            retrieval_contexts=[
+                RetrievedChunksForSpan(
+                    span_id="span ID where data was retrieved",
+                    chunks=[
+                        RetrievedChunk(
+                            content="text content as string (required)",
+                            doc_uri="source URI if available, otherwise null",
+                        )
+                    ],
+                )
+            ]
+        ).model_dump(),
+        indent=2,
+    )
+
+    model_uri = model or get_default_model()
+
+    # Skip LLM extraction if using Databricks default model
+    #
+    # TODO: Implement client logic for calling Databricks default model
+    # with tools and structured output for information extraction
+    if model_uri == _DATABRICKS_DEFAULT_JUDGE_MODEL:
+        _logger.debug(
+            f"Skipping LLM-based retrieval extraction: model '{model_uri}' "
+            "does not support tool calling for trace analysis"
+        )
         return {}
 
-    retrieved = {}
+    try:
+        messages = [
+            ChatMessage(
+                role="system",
+                content=(
+                    "You are a trace analysis expert. Your job is to examine agent execution"
+                    " traces, identify spans (operations) where the agent retrieved information"
+                    " (e.g. by querying a knowledge base, searching documents, or calling an API), "
```

Comment:
"Calling an API" seems too general to be used for the Retrieval* metrics. They should be specific to document retrieval.

Eg we describe [RetrievalRelevance](https://mlflow.org/docs/latest/api_reference/python_api/mlflow.genai.html) as "Retrieval relevance measures whether each chunk is relevant to the input request."